### PR TITLE
CARBONDATA-748

### PR DESCRIPTION
**Pull request by simafengyun**
fix jira CARBONDATA-748, use binary search to replace linear search
the performance is much better no ,from more than 10 seconds to a few millisecond

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Can one of the admins verify this patch?

File: core/src/main/java/org/apache/carbondata/core/util/ByteUtil.java [hunk 1]
```diff
@@ -152,6 +152,159 @@ static boolean lessThanUnsigned(long x1, long x2) {
       return (x1 + Long.MIN_VALUE) < (x2 + Long.MIN_VALUE);
     }
 
+
```
Comment (Hexiaoqiao) [inline core/src/main/java/org/apache/carbondata/core/util/ByteUtil.java]: please delete this annotation if not use any more.

File: core/src/main/java/org/apache/carbondata/core/util/ByteUtil.java [hunk 2]
```diff
@@ -152,6 +152,159 @@ static boolean lessThanUnsigned(long x1, long x2) {
...
+			}
+			return -(low + 1); // key not found.
+		}*/
+	/**
+	 * Checks that {@code fromIndex} and {@code toIndex} are in the range and toIndex % keyLength = 0
+	 * and throws an exception if they aren't.
+	 */
+	private static void rangeCheck(int arrayLength, int fromIndex, int toIndex, int keyWordLength) {
+		if (fromIndex > toIndex || toIndex % keyWordLength != 0) {
+			throw new IllegalArgumentException("fromIndex(" + fromIndex + ") > toIndex(" + toIndex + ")");
```
Comment (Hexiaoqiao) [inline core/src/main/java/org/apache/carbondata/core/util/ByteUtil.java]: the exception msg may be misleading, please give some msg about `toIndex % keyWordLength != 0`.

File: core/src/main/java/org/apache/carbondata/scan/filter/executer/IncludeFilterExecuterImpl.java [hunk 3]
```diff
@@ -154,24 +154,45 @@ private BitSet setFilterdIndexToBitSetWithColumnIndex(
...
+	      for (int k = 0; k < filterValues.length; k++) {
+	    	  
+	    	  int[] index = ByteUtil.UnsafeComparer.INSTANCE.binaryRangeSearch(dataChunk, 0, dataChunk.length, filterValues[k]);
+	    	  for(int i = index[0]; i<=index[1];i++){
+	    		  
+	    		  bitSet.set(i);
+	    	  }
+	      }
+	  
+	      //below comments code is used to confirm the binary range search is correct
```
Comment (Hexiaoqiao) [inline core/src/main/java/org/apache/carbondata/scan/filter/executer/IncludeFilterExecuterImpl.java]: please delete annotation if not used.

File: core/src/main/java/org/apache/carbondata/core/util/ByteUtil.java [hunk 4]
```diff
@@ -152,6 +152,159 @@ static boolean lessThanUnsigned(long x1, long x2) {
...
+  /**
+   * use to search a specific keyword's lower boundary and upper boundary according to upFindFlg
+   * @param dataChunk
+   * @param fromIndex
+   * @param toIndex
+   * @param keyWord
+   * @param upFindFlg true:find upper boundary  false: find lower boundary
+   * @return boundary index 
+   */
+	public static int binaryRangeBoundarySearch(byte[] dataChunk, int fromIndex, int toIndex, byte[] keyWord, boolean upFindFlg) {
```
Comment (Hexiaoqiao) [inline core/src/main/java/org/apache/carbondata/core/util/ByteUtil.java]: 1. there are some common codes between `binaryRangeSearch` and `binaryRangeBoundarySearch`, is it possible to merge them into one method?
2. are you interesting to impl. a complete function about Interval (of Mathematics).
FYI.

**Comment by Hexiaoqiao** [comment]
@simafengyun 
thanks for your work, please rename title follow the format `[CARBONDATA-issue number>] Title of the pull request`.

**Comment by zzcclp** [comment]
Is this pr just be fixed for branch-0.2?

**Comment by ravipesala** [comment]
add to whitelist

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1022/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1023/

**Comment by simafengyun** [comment]
yes
I just fixed on branch-0.2.
If code is ok, I can continue to fix for version 1.0










At 2017-03-07 17:20:44, "Zhichao Zhang" <notifications@github.com> wrote:


Is this pr just be fixed for branch-0.2?

**Comment by ravipesala** [comment]
@simafengyun Thanks for working on it.
Range binary search is impressive. we can use same search in other places as well.
But using this search in `setFilterdIndexToBitSet` is not always correct. Because here the data may not be sorted always, Only for the first column of dimensions is sorted naturally because of mdk order.
And if the data is sorted explicitly(it means it has inverted index) then it goes to another method `setFilterdIndexToBitSetWithColumnIndex`.  So here we need extra checks to do binary search in the method `setFilterdIndexToBitSet` , that is like whether the column is naturally sorted or not.This information we may need get from store.  
Please use this Range binary search in 'setFilterdIndexToBitSetWithColumnIndex' as well.

**Comment by ravipesala** [comment]
@simafengyun Please raise this PR on master as we don't have any plans to provide patches on branch-0.2. And also please format the code.

**Comment by simafengyun** [comment]
You mentioned the below,
>>>>>>>>>>>>>>>>>>>>>>>>>>
But using this search in setFilterdIndexToBitSet is not always correct. Because here the data may not be sorted always, Only for the first column of dimensions is sorted naturally because of mdk order.
>>>>>>>>>>>>>>>>>>>>>>>>>>
 but i don't think so. 
the order you mentioned is the logical level order(MDK).
As I know, for dimension column, it has physical order in chunk level.
for dimension data which has dictionary encode, the dictionary data will sorted in blocklet level and keep the order in chunk on the physical  disk.
So after the one chunk dimension data read, it will keep the order, so I think it is fit for the binary search.


if I was wrong, please feel free to tell me, thanks 










At 2017-03-07 17:44:28, "Ravindra Pesala" <notifications@github.com> wrote:


@simafengyun Thanks for working on it.
Range binary search is impressive. we can use same search in other places as well.
But using this search in setFilterdIndexToBitSet is not always correct. Because here the data may not be sorted always, Only for the first column of dimensions is sorted naturally because of mdk order.
And if the data is sorted explicitly(it means it has inverted index) then it goes to another method setFilterdIndexToBitSetWithColumnIndex. So here we need extra checks to do binary search in the method setFilterdIndexToBitSet , that is like whether the column is naturally sorted or not.This information we may need get from store.
Please use this Range binary search in 'setFilterdIndexToBitSetWithColumnIndex' as well.

**Comment by ravipesala** [comment]
@simafengyun you are right, default we sort the data but we can avoid sorting using `no_inverted_index` property at tblproperties level. So if user uses this configuration then we cannot do binary search always.

**Comment by simafengyun** [comment]
so binary search only fit the below cases，please comfirm thanks

1 the first dimension column ， since it's logicl order are the same with physical order in blocklet level，so it does not need inverted index

2 for dimension column which has inverted index

question 1

if a dimension column is not the first colmn and in a blocklet it just has the same order for both logical and physical，will carbondata create inverted index for it?

question 2

is there an easy way to identify whether a dimension column has no_inverted_index property config by user in source code ，so it will be helpful for me to add a condition to use binary search



thanks

在 2017-03-07 18:35:05，"Ravindra Pesala" <notifications@github.com> 写道：


@simafengyun you are right, default we sort the data but we can avoid sorting using no_inverted_index property at tblproperties level. So if user uses this configuration then we cannot do binary search always.

**Comment by ravipesala** [comment]
Yes, those are the 2 cases we can do binary search on dimension.

Regarding question1 , no it does not create inverted index if the data is already sorted(naturally sorted).

Regarding question2, it is an issue pending from long time, it requires changes from data writer and as well as data reader to update the sort state.  
There are attributes like SortState and Encoders in DataChunk of format needs to updated properly while writing and get them back while reading to know the dimension has inverted_index or not, or it is naturally sorted or not. 

My suggestion is don't change the method of  `setFilterdIndexToBitSet` as the old logic is required for no_inverted index and non sorted columns. Update the the method `setFilterdIndexToBitSetWithColumnIndex` with your range search.  To know the dimension is naturally sorted or not we can add another PR to update the read/write logic.

**Pull request closed without merging**